rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,286 @@
+
+- Leaving a JWT unsigned, and setting the algorithm header to `none` means that JWT validation will succeed in libraries that have not been designed to prevent this flaw.
+- In cases where a service uses asymmetric algorithms for JWT signing, some libraries are vulnerable to accepting a JWT that has been signed used a symmetric algorithm, with the public key of the issuer used as the pre-shared key.
+
+By enforcing an allow-list (to only common battle-tested asymmetric algorithms) of algorithms, and checking this list as part of JWT validation, we mitigate these two vulnerabilities.
+
+Configuring an allow-list also allows us to remove algorithms if a vulnerability is discovered in a specific one.
+
+#### JWT Exfiltration
+
+This broadly falls into two categories:
+
+The first involves a malicious actor intercepting the JWT sent by the node during the join process. Whilst TLS prevents this in most cases, a malicious actor with sufficient access to the environment would be able to insert their own CA into the root store and use this to man-in-the-middle the connection.
+
+The second involves a malicious actor with direct access to executing code within the workload environment. Here they can follow the same steps the node would take in obtaining the JWT.
+
+In both cases, the malicious actor can then use this JWT to gain Teleport credentials and access to customer infrastructure.
+
+Mitigations:
+
+Firstly, we should ensure in documentation that we stress that an attacker with access to the workload environment will be able to access resources that the workload would have access to. The principle of least privilege should be encouraged, with workloads given access to only the resources they require.
+
+In most cases, it appears the providers issue these workload identity JWTs with relatively short lives of between 5 and 10 minutes (GHA and GCP respectively). This does limit a malicious actors ability to continually use these credentials if they lose access to the environment in which they are generated, but, if the attacker exchanges these for a long-lived set of Teleport credentials then this point is relatively moot. We can further limit their access by reducing the TTLs of certificates we offer in exchange for OIDC JWTs, and preventing renewals: requiring a fresh JWT from the environment each time they want to generate certificates. This would be a significant change to the current model whereby Teleport nodes receive long lived certificates (10/infinity years), and present complications in how we currently handle CA rotations. This does feel like the most effective mitigation, as it requires the malicious actor maintain access to the compromised workload environment for roughly as long as they want to access infrastructure.
```

Comment:
Yeah, here I'm mostly referring to standard node joining rather than `tbot`.

I believe that once https://github.com/gravitational/teleport/issues/16691 is fixed, this isn't really a concern on the Machine ID front as we already support users setting this TTL to a value they are comfortable with. But we should really get around to giving shorter-lived certificates to nodes in the future.